contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,193 @@
+# topology aware routing of services
+
+Author: @m1093782566
+
+## Objective
+
+Provide a generic way to provide the local-only service route, say "topology aware routing of service". Local means "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc. To clarify, this proposal is intended to resolve the user case that user always want to visit a local service's backend even though for some reason local service's backend is not available.
+
+## GOAL
+
+A generic way to support various topology levels, regarless whatever the topology key is.
+
+## Non-goal
+
+Scheduler spreading to implement this sort of topology guarantee.
+
+## Use cases
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing service. For example, want to visit a "same-zone" service,  of course, a "same-host" backend is more preferred.
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
```

Comment:
It might be good to describe (precisely) before diving too far into the details:
- How locality is defined. Is it in terms of physical things (nodes, etc), user defined (labels on nodes), something else?
- At what level the routing will occur:
  - Returned by endpoints object in k8s API?
  - DNS entries?
  - Via an L3/4 VIP such as kube proxy?
- What kinds of routing policy will be supported and what the decision algorithm looks like.

Reply:
It would be ideal for the solution to be layerable, like how endpoints are now, usable directly from k8s API, via DNS and integrated with a LB (i.e. kube proxy). But the user is able to chose how to hook it into their setup.

Reply:
> How locality is defined. Is it in terms of physical things (nodes, etc), user defined (labels on nodes), something else?

locality is defined by user, it can be any topology-related thing. "Local" can be "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc.

As to "physical things(nodes)" or  "user-defined thing(labels on nodes)", I think it's "user-defined thing". When we say topologyKey "hostname", we expect every node carries different node labels on "hostname".

Reply:
> At what level the routing will occur:
> * Returned by endpoints object in k8s API?
> * DNS entries?
> * Via an L3/4 VIP such as kube proxy?

routing decision is expected to be implemented in L3/4 VIP level such as kube proxy. Thanks! That's a good point.

Reply:
> What kinds of routing policy will be supported and what the decision algorithm looks like.

If multiple endpoints satisfy the "hard" topology requirement, we will evaluate "how much preferred" each endpoint is by the "soft" topology requirement, and implement something like "weighted round robin".

If no endpoints meet the "hard" topology requirement, simply return "no available service backends" error.